src/screencast/wlr_screencast.c

```diff
@@ -133,6 +134,9 @@ static void wlr_frame_buffer_done(void *data,
 	struct xdpw_screencast_instance *cast = data;
 
 	logprint(TRACE, "wlroots: buffer_done event handler");
+	
+	fps_limit_apply(&cast->fps_limit, cast->max_fps);
```

Comment:
Because xdpw is single-threaded and `nanosleep` ties up our only thread, calling this function here will delay all of xdpw. This probably doesn't create any issues for us, but it would be better if we could conditionally wait to request another frame in the event loop.

This could be addressed by queueing the call to reregister the frame listener callback ([currently invoked here](https://github.com/emersion/xdg-desktop-portal-wlr/blob/e103e120e20e83acec5b66ce4cd33a0eb86df0b0/src/screencast/wlr_screencast.c#L54)) and invoking the next calls in that queue [somewhere in the event loop](https://github.com/emersion/xdg-desktop-portal-wlr/blob/e103e120e20e83acec5b66ce4cd33a0eb86df0b0/src/core/main.c#L175-L217) after checking that the required amount of time has passed.

I'm curious to hear @emersion's thoughts as to whether what I'm suggesting is overkill or not.

Reply:
Yes, I agree we should do our best not to block anywhere.

For reference, mako's event loop allows registering timers: https://github.com/emersion/mako/blob/dbd9c2bc98ef0cec7801a049f4bf8d61f7be36d0/event-loop.c#L128

Reply:
How do you want to proceed? Introducing support for this in the event loop and registering callbacks looks like a major change. Do you, maintainers, want to work on this, or should I try to implement it? My C skills are rusty but I think I might be able to figure it out, though I cannot make any promises.

Reply:
Your C looks at least as good as mine. I'd be fine with you taking a crack at it if you are willing. I'll do my best to review as you go and @emersion has always been super helpful with my crappy C.

You may also want to look at https://github.com/any1/aml, which is the eventloop wayvnc/neatvnc uses. I pointed you to that code base once before, so maybe you're already more familiar with it.

I'm not sure we want to bother using aml as a dependency, but maybe? Should be easy enough to register timers without it, but it may still be a good reference for how to do it.

Reply:
When taking a look at wayvnc, I've noticed that it uses https://github.com/any1/aml as its basic event loop mechanism, including using it for timing the next frame capture. I was basically hoping that this sleep-based solution in this PR would be good enough for `xdg-desktop-portal-wlr` :P

With regards to the technical implementation, which option do you prefer, @danshick @emersion ? Should I go with integrating `aml` into this codebase, which would replace its exisitng event loop, or should I extend the existing event loop with support for timers, based on something like `mako`? The latter makes more sense to me, since its simpler and both `mako` and `xdg-desktop-portal-wlr` are @emersion 's projects. I would just like to confirm before spending too much time with this.

Reply:
Yeah, I think following mako's example makes the most sense. I only shared the link to aml as an alternative example. I don't think we need that level of capability here.

Reply:
Yeah, I've had a look at aml a while ago. My main gripes with it are:

- It involves threading, which always makes things more complicated. The workers could just have been left out, it's easy for the caller to use a pipe instead.
- It involves reference counting for all of its data structures, which is pretty unnecessary for an event loop. Also has a global registry of object IDs for some reason.
- Lack of type safety, many functions take a `void *`.

I still haven't found any go-to event loop library. The Wayland server event loop is pretty good, but I'm not too comfortable depending on wayland-server in a client.

tl;dr is that I've been including a mini event loop in each client for now, with just enough features for the client's purposes.

Reply:
Added a new commit that introduces timers based on [emersion/mako](https://github.com/emersion/mako)'s timer implementation, mostly copy-pasting the changes required for the definition of timers and the event loop, but changing the `mako` prefix to `xdpw`. Also changed the resolution, from milliseconds to nanoseconds.

Also adapted the rate control code now to the new method, which involved some refactoring as well. Retested the actual rate limit, and I think it is more precise now. 

I did not manage to execute the suggested test of running xdpw with a capture configured to a reallt low frame-rate (e.g. `0.1` FPS) and doing a screenshot in the meantime, because the link to `trigger-screen-shot.py` in the [Contributing](https://github.com/emersion/xdg-desktop-portal-wlr/blob/master/CONTRIBUTING.md) guide is now dead. Got no luck with [portal-test](https://github.com/matthiasclasen/portal-test) either, even though I managed to build it locally, it was running buggy and, for example, could not do two subsequent screenshots with it (triggered errors in xdpw). Does anybody still have `trigger-screen-shot.py` somewhere?

Reply:
I may have it lying around on a local dev machine. I'll see if I can dig it up.

You might also try [ksnip](https://github.com/ksnip/ksnip). Not sure how easy it is to explicitly specify portal screenshot as the technique, I've never used it.

Reply:
I managed to test this with ksnip. It worked well. Based on the trace logs, I could see that knsip was able to take the screen shot using my `xdgw` instance while otherwise it was "waiting" between two consecutive frames for screen sharing (10 seconds with FPS = 0.1).

Reply:
I'll mark this thread closed since this was implemented. Please reopen if you have further comments on this subject.